### PR TITLE
Extended ID permissions supported by bidder

### PR DESCRIPTION
**Pull request by YerkovichM**
## Type of change

- [x] Feature

## Description of change

This is reopened https://github.com/prebid/Prebid.js/pull/6112 pull request 
It got reverted because of the bug https://github.com/prebid/Prebid.js/pull/6269

#5814 
Bid adapters receive only allowed user ids

**Case 1**: 
```
userSync: {
 userIds: [
  {
    name: "sharedId",
    bidders: [
      'rubicon',
      'exBidder'
    ],
    params: {
      syncTime: 60 //inseconds,
      defaultis24hours
    },
    storage: {
      type: "cookie",
      name: "sharedid",
      expires: 28
    }
  }
]
}
```

Rubicon bid adapter receives:
```
{
  "bidder": "rubicon",
  ...
  "userId": {
    "sharedid": {
      "id": "01EQDMBY1YPFFVY57R2V11PY89",
      "third": "01EQDMBY1YPFFVY57R2V11PY89"
    }
  },
  "userIdAsEids": [
    {
      "source": "sharedid.org",
      "uids": [
        {
          "id": "01EQDMBY1YPFFVY57R2V11PY89",
          "atype": 1,
          "ext": {
            "third": "01EQDMBY1YPFFVY57R2V11PY89"
          }
        }
      ]
    }
  ],
  ...
}
```
PrebidServerBidAdapter request sent as:
```
{
  ...
  "ext": {
    "prebid": {
      ...
      "data": {
        "eidpermissions": [
          {
            "source": "sharedid.org",
            "bidders": [
              "rubicon"
            ]
          }
        ]
      }
    }
  },
  ...
  "user": {
    "ext": {
      "eids": [
        {
          "source": "sharedid.org",
          "uids": [
            {
              "id": "01EQDMBY1YPFFVY57R2V11PY89",
              "atype": 1,
              "ext": {
                "third": "01EQDMBY1YPFFVY57R2V11PY89"
              }
            }
          ]
        }
      ]
    }
  }
}
```
**Case 2**:
```
userSync: {
  userIds: [{
    name: "sharedId",
    bidders: ['exBidder'],
    ...
    }
  ]
}
```
Rubicon bid adapter receives:
```
{
  "bidder": "rubicon",
  ... no user ID entries ...
}
```
PrebidServerBidAdapter request:
```
{
  ...
  "ext": {
    "prebid": {
      ...
      "data": {
        "eidpermissions": [
          {
            "source": "sharedid.org",
            "bidders": [ 'exBidder' ]
          }
        ]
      }
    }
  },
  ...
}
```
**Case 3**: no permissions specified
```
userSync: {
  userIds: [{
    name: "sharedId",
    ...
    }
  ]
}
```
Rubicon bid adapter receives:
```
{
  "bidder": "rubicon",
  ...
  "userId": {
    "sharedid": {
      "id": "01EQDMBY1YPFFVY57R2V11PY89",
      "third": "01EQDMBY1YPFFVY57R2V11PY89"
    }
  },
  "userIdAsEids": [
    {
      "source": "sharedid.org",
      "uids": [
        {
          "id": "01EQDMBY1YPFFVY57R2V11PY89",
          "atype": 1,
          "ext": {
            "third": "01EQDMBY1YPFFVY57R2V11PY89"
          }
        }
      ]
    }
  ],
 ...
}
```
PrebidServerBidAdapter request:
```
{
  ...
  "ext": {
    "prebid": {
      ...
      "data": {
         ...
      }
    }
  },
  ...
  "user": {
    "ext": {
      "eids": [
        {
          "source": "sharedid.org",
          "uids": [
            {
              "id": "01EQDMBY1YPFFVY57R2V11PY89",
              "atype": 1,
              "ext": {
                "third": "01EQDMBY1YPFFVY57R2V11PY89"
              }
            }
          ]
        }
      ]
    }
  }
}
```
@jdwieland8282

### CODE REVIEW

**Comment by bretg** [comment]
@YerkovichM - how did you fix the dependency between pbsBidAdapter and the eidPermissions function? Doesn't look like any changes to me.

@jsnellbaker, can you recommend an approach?

**Comment by YerkovichM** [comment]
@bretg thought issue with the dependency will be topic of the discussion here.

Here what I propose: 
We can make a setter function globally from prebidServerBidAdapter for userId module to use.
When userId module initialize its submodules it will set eidPermission via global setter.
There will be no file or module dependency between prebidServerBidAdapter and userId module, but we should make sure that setter function is not used by other modules.
`getGlobal().setEidPermissions` 
This is implemented in last commit

**Comment by jsnellbaker** [comment]
@YerkovichM Thanks for the updates & comments.  

I think the overall approach you've outlined and implemented through the recent commits are what we need to do in this case.  My only potential concern is using `getGlobal()`.  

This is the public API object for Prebid, so setting this setter function here will make it externally available for use.  I'm not sure this function is something we'd want to expose?  If not, it may be more ideal to try to house the setter function through the `utils` file.

Please let me know your thoughts on the above.

**Comment by YerkovichM** [comment]
@jsnellbaker
Implemented feature through the utils file
Let me know what you think

File: plugins/eslint/validateImports.js [hunk 1]
```diff
@@ -26,7 +26,8 @@ function flagErrors(context, node, importPath) {
     if (
       path.dirname(absImportPath) === absModulePath || (
         absImportPath.startsWith(absModulePath) &&
-        path.basename(absImportPath) === 'index.js'
+        path.basename(absImportPath) === 'index.js' &&
+        path.basename(absFileDir) !== 'prebidServerBidAdapter'
```
Comment (jsnellbaker) [inline plugins/eslint/validateImports.js:30]: Is this still needed anymore, since we changed how the imports work for the PBS adapter?

Comment (YerkovichM) [reply plugins/eslint/validateImports.js]: No need for this anymore
Removed changes

**Review by jsnellbaker — APPROVED** [review]
I think this looks fine now.

Just need a 2nd review.

**Review by harpere — APPROVED** [review]
LGTM

**Pull request merged**